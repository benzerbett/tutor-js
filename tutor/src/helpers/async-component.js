import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
