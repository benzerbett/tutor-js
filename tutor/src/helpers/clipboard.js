import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
