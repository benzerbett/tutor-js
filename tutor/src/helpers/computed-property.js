import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
