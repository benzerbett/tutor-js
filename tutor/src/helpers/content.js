import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
