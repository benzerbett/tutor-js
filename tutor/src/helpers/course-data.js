import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
