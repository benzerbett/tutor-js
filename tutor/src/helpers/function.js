export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
