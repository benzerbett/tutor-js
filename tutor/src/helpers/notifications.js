import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
