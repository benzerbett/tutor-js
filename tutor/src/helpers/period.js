import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
