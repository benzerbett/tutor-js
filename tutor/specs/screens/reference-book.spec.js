import { React, SnapShot } from '../helpers';
import { Promise } from 'es6-promise';
import Factory, { FactoryBot } from '../factories';

import Book from '../../src/models/reference-book';
import Page from '../../src/models/reference-book/page';

import Router from '../../src/helpers/router';
import ReferenceBook from '../../src/screens/reference-book/reference-book';
import { bootstrapCoursesList } from '../courses-test-data';
import ReferenceBookUX from '../../src/screens/reference-book/ux';
import EnzymeContext from '../helpers/enzyme-context';

jest.mock('../../src/helpers/router');

const COURSE_ID = '1';

describe('Reference Book Component', function() {
  let props, ux, course, REFERENCE_BOOK, REFERENCE_BOOK_PAGE_DATA, router;

  beforeEach(function() {
    router = { history: { push: jest.fn() } };
    ux = new ReferenceBookUX(router);
    ux.book = Factory.book();
    ux.setChapterSection('2.1');
    props = {
      ux,
    };
    Router.currentParams.mockReturnValue({
      ecosystemId: ux.book.id,
      chapterSection: ux.page.chapter_section.toString,
    });

  });

  it('renders the section title on the navbar', () => {
    const book = mount(<ReferenceBook {...props} />, EnzymeContext.build());
    expect(book).toHaveRendered('BookPage');
  });

  it('renders page html that matches snapshot', function() {
    const book = mount(<ReferenceBook {...props} />, EnzymeContext.build());
    expect(book.html()).toMatchSnapshot();
  });

  it('navigates forward and back between pages', function() {
    expect(ux.page.chapter_section.asString).toEqual('2.1');
    Router.makePathname.mockReturnValue('/test/route/3');
    const book = mount(<ReferenceBook {...props} />, EnzymeContext.build());
    book.find('.paging-control.next').simulate('click');
    expect(router.history.push).toHaveBeenCalledWith('/test/route/3');
  });


  it('sets the menu item to be active based on the current page', function() {
    const book = mount(<ReferenceBook {...props} />, EnzymeContext.build());
    expect(book).toHaveRendered(`.book-menu [data-section='${ux.page.chapter_section.asString}'] .active`);
  });

  it('displays a not found message when needed', () => {
    ux.chapterSection = '99.99';
    expect(ux.isFetching).toBe(false);
    const book = mount(<ReferenceBook {...props} />, EnzymeContext.build());
    expect(book).toHaveRendered('.not-found');
    expect(book.text()).toContain('Section 99.99 was not found');
  });

});
