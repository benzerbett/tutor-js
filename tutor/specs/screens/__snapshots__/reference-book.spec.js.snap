// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference Book Component renders page html that matches snapshot 1`] = `"<div class=\\"reference-book menu-on-top\\"><div class=\\"openstax-debug-content\\"><div class=\\"obscured-page\\"><div class=\\"page obscured-page__Page-c4o3c2-1 eeKcmC\\"><div class=\\"content\\"><div class=\\"book-menu ontop\\"><ul class=\\"section\\" data-depth=\\"1\\"><li data-section=\\"1\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">1</span>Kinematics</a></li><li data-section=\\"1.1\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"1.1\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">1.1</span>Kinematics in 1 dim</a></li></ul></li><li data-section=\\"1.2\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"1.2\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">1.2</span>Kinematics in 2 dim</a></li></ul></li><li data-section=\\"1.3\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"1.3\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">1.3</span>Kinematics in 3 dim</a></li></ul></li></ul><ul class=\\"section\\" data-depth=\\"1\\"><li data-section=\\"2\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">2</span>Mechanics</a></li><li data-section=\\"2.1\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"2.1\\"><a tabindex=\\"-1\\" class=\\"active\\"><span class=\\"section-number\\">2.1</span>Mechanics in 1 dim</a></li></ul></li><li data-section=\\"2.2\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"2.2\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">2.2</span>Mechanics in 2 dim</a></li></ul></li><li data-section=\\"2.3\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"2.3\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">2.3</span>Mechanics in 3 dim</a></li></ul></li></ul><ul class=\\"section\\" data-depth=\\"1\\"><li data-section=\\"3\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">3</span>Thermodynamics</a></li><li data-section=\\"3.1\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"3.1\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">3.1</span>Thermodynamics in 1 dim</a></li></ul></li><li data-section=\\"3.2\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"3.2\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">3.2</span>Thermodynamics in 2 dim</a></li></ul></li><li data-section=\\"3.2\\"><ul class=\\"section\\" data-depth=\\"2\\"><li data-section=\\"3.2\\"><a tabindex=\\"-1\\" class=\\"\\"><span class=\\"section-number\\">3.2</span>Thermodynamics in 3 dim</a></li></ul></li></ul></div><div class=\\"tutor-paging-navigation book-page-wrapper\\"><a href=\\"#\\" target=\\"_blank\\" tabindex=\\"0\\" title=\\"Go Back\\" aria-controls=\\"paged-content\\" class=\\"paging-control prev\\"><div class=\\"arrow-wrapper\\"><svg class=\\"icon arrow left\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0px\\" y=\\"0px\\" viewBox=\\"0 0 370 735\\"><path d=\\"M369.225,658.604l0,76.396l-369.225,-369.225l0.3,-0.3l368.925,-365.475l0,77.168l-291.449,290.332l291.449,291.104Z\\"></path></svg></div></a><div tabindex=\\"0\\" role=\\"region\\" aria-live=\\"polite\\" id=\\"paged-content\\" class=\\"paged-content\\"><div class=\\"book-page\\"><div class=\\"page center-panel\\"><h4 class=\\"related-content\\"><span class=\\"part\\"><span class=\\"section\\">2.1 </span><span class=\\"title\\">Mechanics in 1 dim</span></span></h4><div class=\\"openstax-has-html book-content\\"></div></div><div class=\\"visible-when-debugging ecosystem-info\\">PageId: 1e58aa87-2e09-40a7-8bf3-269b2fa16511, Ecosystem: </div></div></div><a href=\\"#\\" tabindex=\\"0\\" title=\\"Go Forward\\" aria-controls=\\"paged-content\\" class=\\"paging-control next\\"><div class=\\"arrow-wrapper\\"><svg class=\\"icon arrow right\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0px\\" y=\\"0px\\" viewBox=\\"0 0 370 735\\"><path d=\\"M0,658.604l0,76.396l369.225,-369.225l-0.3,-0.3l-368.925,-365.475l0,77.168l291.449,290.332l-291.449,291.104Z\\"></path></svg></div></a></div></div></div><div class=\\"overlay obscured-page__Overlay-c4o3c2-0 eitCT\\"></div></div><a href=\\"#spy\\" tabindex=\\"-1\\" class=\\"debug-toggle-link\\" aria-hidden=\\"true\\">Ï€</a></div></div>"`;
